### ISSUE TITLE
Spaces: API attribute "sync-local-parent" to keep locally synced Spaces in a common parent folder (Desktop)

### ISSUE THREAD

**Issue by michaelstingl** [owncloud/ocis#2794]
## Is your feature request related to a problem? Please describe.

@hodyroff brought up, regular users expect to sync [all Spaces to local disk by default](https://github.com/owncloud/ocis/issues/2793), and Spaces should be placed in a local "`Spaces`" parent folder.

```
.
├── Home
├── Shares
│   ├── FromAlice
│   └── FromBob
└── Spaces
    ├── Space1
    ├── Space2
    └── Space3
```

## Describe the solution you'd like

Flexible solution would be, Spaces API could get an additional attribute "sync-local-parent" to control the default behaviour of the desktop sync client.

In the first iteration, this attribute could be hardcoded – no backend logic needed.

In the future, we could also support scenarios with all Spaces in account root:

```
.
├── Home
├── Shares
│   ├── FromAlice
│   └── FromBob
├── Space1
├── Space2
└── Space3
```

…or use different terminology:

```
.
├── Home
├── Shares
│   ├── FromAlice
│   └── FromBob
└── Project
    ├── Project1
    ├── Project2
    └── Project3
```

## Describe alternatives you've considered

- Not place all Spaces in a common folder "Spaces"
- Hardcode in desktop sync client to place all Spaces in a common folder "Spaces"

**Comment by refs** [comment]
> regular users expect to sync all Spaces to local disk by default

Interesting; do we have user reports already for spaces? that support this? Which type of user expect this behavior?

**Comment by butonic** [comment]
The spaces concept shifts the responsibility of determining the local mount path for a space from the server to the clients. In oc10 the server is responsible for mounting a (federated) share in the users home. When dealing with spaces there is no longer a global (human readable) namespace, because every space has an id and a URL and no central server can know all mount paths.

The `/me/drives` endpoint serves as a list of spaces that a user has access to. It is provided by a space registry.

@micbar and I are having a discussion about whether or not a registry persists the mapping of mount points or aliases to spaces. IMO our understanding of the concepts diverges as follows:

1. In my opinion aliases are mount points, not just another property of a space. Spaces already have a name and a description. They can be further distinguished by the space image, the owner (if present), or the domain that serves the space. While this alon should allow clients to present the spaces in a meaningful way to end users there is a challenge for desktop clients, which relates to this issue: they need to make all spaces locally accessible, which means building a local namespace that might include spaces with the same name. When such a name clash occurs, eg. two spaces have the same name `Documents` the client needs to use find two different names. In oc10 the server would just append a ` (1), (2), ...` to the name. That is what users are used to, now (not only from ownCloud but also other software). While URLs always contain a spaceid suffix, we also want to have an space alias in the url to have human readable urls. When there is no name collision the alias might be the name of the space, or `{spacetype}/{spacename}` to distinguish different types of spaces. But space names are not unique and clients must be able to choose a different alias for the space then. Ultimately, end users will want to set their own alias for a space and eg. rename `Projects/Documents (3)` to `Projects/Marketing Documents`.

2. Space aliases are user specific: when client a aliases a space as `Projects/Documents (3)` it should have the same name on another client. But only for the same user. The registry / the `/me/drives` endpoint is used to let clients discover an alias and present the same local namespace to the end user, regardless of the device they are using.

3. The space registry is responsible for persisting the list of spaces a user has access to as well as the aliases aka mount points for them. It can then poll them on behalf of clients to aggregate etag updates and present a single `/me/drives` listing. Future improvements can also use push notifications to further reduce sync latency. Our current spaces registry is a cache that knows all storage providers and polls them to find all spaces. In a truly federated system the registry must persist the spaces itself because it cannot possibly discover spaces all over the world. When a user gains access to a space a registry has to receive an explicit call to be made aware of that space. We currently just omit that by polling the stroage providers on every request. When there is an (openid connect like) discovery mechanism for the registry (which basically is the ocis instance a user uses to manage his spaces) granting users access to a space should discover the registry of the user and 'register' the space for the user.

**Comment by micbar** [comment]
After a long talk with @dragotin  I still think this is out of scope for GA and the local mount points should be decided by clients. Having a space registry which stores user specific mount points on the server is IMO an antipattern for oCIS

1) oCIS and the spaces concept is decentralized.
2) The server should not be responsible for storing client-  AND user-specific information
3) Clients can be connected to multiple ocis instances, so a central space registry is not desired.
4) Storing user-specific information about spaces  in the registry would introduce a new split-brain situation.

**Comment by dragotin** [comment]
What @micbar says. Plus: The clients need to be self contained in case the server is not reachable.

Later we can build the functionality of saving user settings to the server for persistence reasons, but than it is an attribute of (user, client) and not the space.

**Comment by butonic** [comment]
> After a long talk with @dragotin I still think this is out of scope for GA and the local mount points should be decided by clients. Having a space registry which stores user specific mount points on the server is IMO an antipattern for oCIS
> 
> 1. oCIS and the spaces concept is decentralized.

true, as are registries. a client can know about multiple registries, registries know multiple spaces, and a registry can persist the mapping of user to space for multiple users. In our current implementation that is all implemented by the storage space registry. Not with persistence, but by polling and caching storage providers about the available storage spaces.

> 2. The server should not be responsible for storing client-  AND user-specific information

What server exactly? We have a `/me/drives` endpoint which is mostly fed by the space registry. And we have `/dav/spaces/{spaceid}` and maybe a trash endpoint per space that is handled by storage providers.

What do you see as client specific information? IMO the alias for a space or if you want to keep that seperate, the mount path of a space is client specific information, bit it should be shared over all clients. You don't want 'Documents (2)' to end up as `Documents (3)` when switching from laptop to workstation. The local namespace presented to the user should be the same. Which means clients need a server to exchange the mount paths. IMO via the registry. We could persist that as a per user individual property on a space root, but then spaces would in effect know where they are mounted and space providers could collect information about all aliases users gave a space ... the registry is responsible to persist the alias. In the end I may want to run my own personal registry and register spaces from my employee, an insurance company and some of my friends. without any of them knowing what I called their space.

I'm not saying that is how it should be right now. We could hardcode a local naming scheme that deterministically builds aliases. I'm open to suggestions. But all of them have problems and ultimately the end user decides (if we allow it ... IMO we should at some point). What are possible schemes that can be used by clients:
- append an increasing number to the end? might produce different ordering, when mounted in different order
- append the creation date? might still collide
- append a portion of the id, eg 8 chars of the hash of the spaceid? might actually work, but what will UX be?
- prefix owner or institution? might still collide

In our current implementation we can enforce a scheme on the server side by preventing the same name for spaces (similar to user names). Or we allow arbitrary names but enforce spaceAlias to be uniqe per instance. Could then guarantee uniqueness for all spaceAliases if we include a human readable instance identifier, eg the domain?

> 3. Clients can be connected to multiple ocis instances, so a central space registry is not desired.

IMO an instance is a registry. Where spaces are hosted does not matter.

> 4. Storing user-specific information about spaces in the registry would introduce a new split-brain situation.

IMO a registry is a highly user specific service. Ultimately, I may want to run a registry on my personal domain and register spaces to it. If a space does not allow registring it in my personal registry I may need to query an additional registry (eg. the one from an employer that wants to enforce a specific namespace).

You might argue that aliases are instance specific. And I would agree that admins want to enforce certain paths, eg. to mimic existing paths in other systems. ;-) I'm happy to allow admins to enforce a namespace with a different registry implementation.  But I also want to be able to use my personal registry ... because I want my clients to only need to ping one registry. Hell, I could even build a registry that aggregates registries.

But if only allow a unique namespace (unique spaceAliases) per instance ... we wouldn't need an id in the URLs because we could build a globally unique namespace by combining instance domain and the unique spaceAlias per instance. But that is not what we were aiming for, right?

@dragotin regarding
> The clients need to be self contained in case the server is not reachable.

What server exactly? The `/me/drives` endpoint of a registry or the `/dav/spaces/{spaceid}` endpoint of a space?

If the `/me/drives` endpoint is not reachable, the client could fall back to a cached verison of the space listing and poll each space individually ... super inefficient, but possible. Maybe a second registry on a different domain can act as a backup registry?

If a space is not reachable ... well, try again later ...

> Later we can build the functionality of saving user settings to the server for persistence reasons, but than it is an attribute of (user, client) and not the space.

What user setting? The mount point?  Then you agree that that is a user specific setting that should be stored in the registry. The space registry could store that in a metadata space, or even a dedicated registry space.

Or an instance unique alias? That would be a space property. @micbar @dragotin wo could combine that and say that a space may have a spaceAlias, eg to enforce a naming scheme by admins. Maybe with a flag to allow / renaming. but a user specific alias would land in the registry.
